components/places/src/ffi.rs

```diff
@@ -110,27 +209,60 @@ fn get_code(err: &Error) -> ErrorCode {
                     // and resolve it.
                     if let Some(places_err) = store_error.downcast_ref::<Error>() {
                         log::info!("Recursing to resolve places error");
-                        get_code(places_err)
+                        get_error_number(places_err)
                     } else {
                         log::error!("Unexpected sync error: {:?}", err);
-                        ErrorCode::new(error_codes::UNEXPECTED)
+                        error_codes::UNEXPECTED
                     }
                 }
                 _ => {
                     // TODO: expose network errors...
                     log::error!("Unexpected sync error: {:?}", err);
-                    ErrorCode::new(error_codes::UNEXPECTED)
+                    error_codes::UNEXPECTED
                 }
             }
         }
 
         err => {
             log::error!("Unexpected error: {:?}", err);
-            ErrorCode::new(error_codes::UNEXPECTED)
+            error_codes::UNEXPECTED
+        }
+    }
+}
+
+/// This is very very hacky - we somehow need to ensure the same error hierarchy
+/// exists for both hand-written FFI functions and those generated by uniffi,
+/// and there doesn't seem to be a clean way of doing that. So our .udl defines
+/// a single error type - ErrorWrapper::Wrapped(). The `String` message there
+/// is, roughly, `format!("{}|{}", extern_error.code, extern_error.message)`.
+/// There then exists code on the Swift and Kotlin side of the world which
+/// unpacks this and returns the exact same error objects as if it was an
+/// `ExternError` in the first place.
```

Comment:
FWIW, I agree with your implied assessment that this hackery should be short-lived enough that it's not worth trying to do anything nicer here 👍🏻